tasks/eva_2389/run_vcf_vertical_concat_pipeline.py

```diff
@@ -0,0 +1,167 @@
+                                            stage=0, prev_stage_processes=[], pipeline=NextFlowPipeline()):
+    """
+    # Generate Nextflow pipeline for multi-stage VCF concatenation of 5 VCF files with 2-VCFs concatenated at a time (CONCAT_CHUNK_SIZE=2)
+    # For illustration purposes only. Usually the CONCAT_CHUNK_SIZE is much higher (ex: 500).
+    #
+    #		    vcf1		    vcf2		vcf3		    vcf4		vcf5
+    #               \		     /		       \		      /
+    # Stage0:	     \		   /		        \		    /
+    # -------	      \	     /		             \	      /
+    #	    	vcf1_2=concat(vcf1,vcf2)	vcf3_4=concat(vcf3,vcf4)	vcf5    <---- 3 batches of concat in stage 0
+    #		 		    \ 		                /
+    # Stage1:	  		 \		              /
+    # -------	   		  \	                /
+    #				vcf1_2_3_4=concat(vcf1_2,vcf3_4)		            vcf5    <---- 2 batches of concat in stage 1
+    #	 					      \ 		                            /
+    # Stage2:	  		 		   \		 	                      /
+    # -------	   		  			\	                            /
+    #						      vcf1_2_3_4_5=concat(vcf1_2_3_4,vcf5)          <----- Final result
+    """
+    if len(vcf_files) == 1: # If we are left with only one file, this means we have reached the last concat stage
+        return pipeline, vcf_files[0]
+    num_batches_in_stage = math.ceil(len(vcf_files) / concat_chunk_size)
+    curr_stage_processes = []
+    output_vcf_files_from_stage = []
+    for batch in range(0, num_batches_in_stage):
+        # split files in the current stage into chunks based on concat_chunk_size
+        files_in_batch = vcf_files[(concat_chunk_size * batch):(concat_chunk_size * (batch + 1))]
+        files_to_concat_list = write_files_to_concat_list(files_in_batch, stage, batch, concat_processing_dir)
+        concat_stage_batch_name = f"concat_stage{stage}_batch{batch}"
+        log_file_name = os.path.join(concat_processing_dir, f"{concat_stage_batch_name}.log")
+        output_vcf_file = get_output_vcf_file_name(stage, batch, concat_processing_dir)
+        process = NextFlowProcess(process_name=concat_stage_batch_name,
+                                  command_to_run=str(PythonProcess(vcf_vertical_concat,
+                                                                   {"files-to-concat-list": files_to_concat_list,
+                                                                    "concat-processing-dir": concat_processing_dir,
+                                                                    "output-vcf-file": output_vcf_file,
+                                                                    "bcftools-binary": bcftools_binary},
+                                                                   log_file=log_file_name
+                                                                   )
+                                                     )
+                                  )
```

Comment:
I don't quite get why we need to kick off a python process in the nextflow rather than just running bcftools concat/index directly from bash... maybe I'm missing something?

Reply:
At this point the concat process is just running "bcftools concat" but we will have to eventually enhance this process because the concat process [does not seem to properly concat the INFO fields](https://www.ebi.ac.uk/panda/jira/browse/EVA-2317?focusedCommentId=353543&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-353543).

Reply:
Using Nextflow makes perfect sense, I was asking about the first part - thanks for the explanation!